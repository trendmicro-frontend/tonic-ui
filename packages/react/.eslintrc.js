module.exports = {
  extends: 'trendmicro',
  parser: '@babel/eslint-parser',
  env: {
    browser: true,
    node: true,
    jest: true,
  },
  plugins: [
    '@babel',
    'react-hooks',
  ],
  rules: {
    'react/prop-types': 0,
    'react-hooks/rules-of-hooks': 'error', // Checks rules of Hooks
    'react-hooks/exhaustive-deps': 'error', // Checks effect dependencies
  },
};
