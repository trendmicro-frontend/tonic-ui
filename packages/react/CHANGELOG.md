# @tonic-ui/react

## 2.0.3

### Patch Changes

- fix(react/OverflowTooltip): resolve tooltip misalignment issue with popup menu items by [@tinaClin](https://github.com/tinaClin) in [#894](https://github.com/trendmicro-frontend/tonic-ui/pull/894)

- feat(react): improve tooltip placement and offset calculations by [@cheton](https://github.com/cheton) in [#897](https://github.com/trendmicro-frontend/tonic-ui/pull/897)
