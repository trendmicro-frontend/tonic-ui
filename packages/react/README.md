React UI components built with Emotion and Styled System
