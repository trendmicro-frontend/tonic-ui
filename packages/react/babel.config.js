module.exports = {
  extends: '@trendmicro/babel-config',
  presets: [
    '@babel/preset-env',
    ['@babel/preset-react', {
      /**
       * Decides which runtime to use.
       * - 'automatic' auto imports the functions that JSX transpiles to.
       * - 'classic' does not automatic import anything.
       */
      'runtime': 'classic', // defaults to classic
      //'runtime': 'automatic', // for React 16.14.0, React 17 and higher that supports the new transform
    }],
    '@emotion/babel-preset-css-prop',
  ],
};
