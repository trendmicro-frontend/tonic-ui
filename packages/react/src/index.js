// components
export * from './accordion';
export * from './alert';
export * from './badge';
export * from './box';
export * from './button';
export * from './checkbox';
export * from './color-mode';
export * from './color-style';
export * from './css-baseline';
export * from './divider';
export * from './drawer';
export * from './flat-button';
export * from './flex';
export * from './grid';
export * from './icon';
export * from './image';
export * from './input';
export * from './link';
export * from './menu';
export * from './modal';
export * from './pagination';
export * from './popover';
export * from './popper';
export * from './portal';
export * from './presence';
export * from './radio';
export * from './search-input';
export * from './scrollbar';
export * from './select';
export * from './skeleton';
export * from './space';
export * from './spinner';
export * from './stack';
export * from './switch';
export * from './table';
export * from './tabs';
export * from './tag';
export * from './text';
export * from './textarea';
export * from './theme';
export * from './toast';
export * from './tooltip';
export * from './transitions';
export * from './visually-hidden';

// hooks
export useDisclosure from './deprecated/useDisclosure'; // deprecated

// settings
export colorStyle from './shared/color-style';
export theme from './shared/theme';
