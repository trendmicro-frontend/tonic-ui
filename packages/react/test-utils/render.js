import '@testing-library/jest-dom/extend-expect';
import { render } from '@testing-library/react';
import { toHaveNoViolations } from 'jest-axe';
import * as React from 'react';
import { TonicProvider, theme } from '../src';

expect.extend(toHaveNoViolations);

const customRender = (ui, options) => {
  let wrapper = TonicProvider;
  const useCSSVariables = options?.useCSSVariables;
  if (useCSSVariables) {
    wrapper = ({ children }) => (
      <TonicProvider
        theme={{
          ...theme,
          config: {
            ...theme.config,
            useCSSVariables: true,
          },
        }}
      >
        {children}
      </TonicProvider>
    );
  }

  return render(ui, { wrapper, ...options });
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
