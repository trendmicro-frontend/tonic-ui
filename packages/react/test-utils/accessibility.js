import { act } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import * as React from 'react';
import { render } from './render';

expect.extend(toHaveNoViolations);

const testA11y = async (ui, options) => {
  const { axeOptions, ...rest } = { ...options };
  const container = React.isValidElement(ui)
    ? render(ui, rest).container
    : ui;
  await act(async () => {
    const results = await axe(container, axeOptions);
    expect(results).toHaveNoViolations();
  });
};

export {
  testA11y,
};
