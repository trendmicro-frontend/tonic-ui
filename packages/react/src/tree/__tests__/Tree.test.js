/* eslint-disable react/jsx-no-bind */
import { act, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { render } from '@tonic-ui/react/test-utils/render';
import { testA11y } from '@tonic-ui/react/test-utils/accessibility';
import {
  Box,
  Flex,
  OverflowTooltip,
  TreeItem,
  TreeItemContent,
  TreeItemToggle,
  TreeItemToggleIcon,
  Tree,
  useColorStyle,
} from '@tonic-ui/react/src';
import { ensureArray } from 'ensure-type';
import React from 'react';

const buildTreeNodes = () => {
  const treeNodes = [
    {
      'id': '1',
      'name': 'Node 1',
      'children': [
        {
          'id': '2',
          'name': 'Node 2',
          'children': [
            {
              'id': '3',
              'name': 'Node 3',
              'children': [
                {
                  'id': '4',
                  'name': 'Node 4',
                }
              ]
            }
          ]
        },
        {
          'id': '5',
          'name': 'Node 5',
          'children': [
            {
              'id': '6',
              'name': 'Node 6',
            }
          ]
        }
      ]
    },
    {
      'id': '7',
      'name': 'Node 7',
      'children': [
        {
          'id': '8',
          'name': 'Node 8',
        }
      ]
    },
    {
      'id': '9',
      'name': 'Node 9',
      'children': [
        {
          'id': '10',
          'name': 'Node 10',
        }
      ]
    }
  ];

  return treeNodes;
};

const buildTreeMap = (nodes) => {
  const treeMap = new Map();

  const traverse = (nodes) => {
    ensureArray(nodes).forEach((node) => {
      treeMap.set(node.id, node);
      traverse(node.children);
    });
  };

  traverse(nodes);

  return treeMap;
};

const findExpandableNodeIds = (nodes) => {
  const expandableNodeIds = [];

  const traverse = (nodes) => {
    ensureArray(nodes).forEach((node) => {
      if (Array.isArray(node.children) && node.children.length > 0) {
        expandableNodeIds.push(node.id);
        traverse(node.children);
      }
    });
  };

  traverse(nodes);

  return expandableNodeIds;
};

const TreeItemRender = ({
  node,
  nodeDepth = 0,
}) => {
  const [colorStyle] = useColorStyle();

  return (
    <TreeItem
      data-testid={`node-${node.id}`}
      key={node.id}
      nodeId={node.id}
      render={({ isExpandable, isSelected }) => { // eslint-disable-line react/jsx-no-bind
        return (
          <TreeItemContent
            data-testid={`node-${node.id}-content`}
            sx={{
              // [Optional] Display a connecting line to indicate which is the last node when hovered over the tree item
              ':hover + [role="group"]': {
                position: 'relative',
                '::before': {
                  backgroundColor: colorStyle.background.highlighted,
                  content: '""',
                  position: 'absolute',
                  top: 0,
                  bottom: 0,
                  left: 20 + nodeDepth * 24 - (1 / 2), // Adjust the horizontal position based on depth
                  width: 1,
                },
              },
            }}
          >
            <Flex
              flex="none"
              width="6x"
            >
              {isExpandable ? (
                <TreeItemToggle
                  aria-label="toggle"
                  data-testid={`node-${node.id}-toggle`}
                >
                  <TreeItemToggleIcon />
                </TreeItemToggle>
              ) : null}
            </Flex>
            <OverflowTooltip label={node.name}>
              {({ ref, style }) => (
                <Box
                  ref={ref}
                  {...style}
                  flex="auto"
                  fontWeight={isSelected ? 'semibold' : 'normal'}
                >
                  {node.name}
                </Box>
              )}
            </OverflowTooltip>
          </TreeItemContent>
        );
      }}
    >
      {ensureArray(node.children).map(node => (
        <TreeItemRender
          key={node.id}
          node={node}
          nodeDepth={nodeDepth + 1}
        />
      ))}
    </TreeItem>
  );
};

describe('Tree', () => {
  it('should render correctly', async () => {
    const renderOptions = {
      useCSSVariables: true,
    };
    const treeNodes = buildTreeNodes();
    const treeMap = buildTreeMap(treeNodes);
    const expandableNodeIds = findExpandableNodeIds(treeNodes);
    const allNodeIds = Array.from(treeMap.keys());

    expect(expandableNodeIds).toStrictEqual(['1', '2', '3', '5', '7', '9']);
    expect(allNodeIds).toStrictEqual(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10']);

    const { container } = render((
      <Tree
        aria-label="tree"
        isSelectable
        isUnselectable
        multiSelect
      >
        {ensureArray(treeNodes).map(node => (
          <TreeItemRender
            key={node.id}
            node={node}
          />
        ))}
      </Tree>
    ), renderOptions);

    expect(screen.getByRole('tree')).toHaveAttribute('aria-multiselectable', 'true');

    await testA11y(container);
  });

  it('should call onBlur when the tree is blurred', () => {
    const handleBlur = jest.fn();
    const treeNodes = buildTreeNodes();

    render(
      <Tree
        onBlur={handleBlur}
      >
        {ensureArray(treeNodes).map(node => (
          <TreeItemRender
            key={node.id}
            node={node}
          />
        ))}
      </Tree>
    );

    act(() => {
      screen.getByRole('tree').focus();
    });

    act(() => {
      screen.getByRole('tree').blur();
    });

    expect(handleBlur).toHaveBeenCalledTimes(1);
  });

  it('should call onFocus when the tree is focused', () => {
    const handleFocus = jest.fn();
    const treeNodes = buildTreeNodes();

    render(
      <Tree
        onFocus={handleFocus}
      >
        {ensureArray(treeNodes).map(node => (
          <TreeItemRender
            key={node.id}
            node={node}
          />
        ))}
      </Tree>
    );

    act(() => {
      screen.getByRole('tree').focus();
    });

    expect(handleFocus).toHaveBeenCalledTimes(1);
  });

  it('should call onKeyDown when a key is pressed', async () => {
    const user = userEvent.setup();
    const handleKeyDown = jest.fn();
    const treeNodes = buildTreeNodes();

    render(
      <Tree
        onKeyDown={handleKeyDown}
      >
        {ensureArray(treeNodes).map(node => (
          <TreeItemRender
            key={node.id}
            node={node}
          />
        ))}
      </Tree>
    );

    act(() => {
      screen.getByRole('tree').focus();
    });

    await user.keyboard('[Enter]');
    await user.keyboard('[ArrowDown]');
    await user.keyboard('[Space]');

    expect(handleKeyDown).toHaveBeenCalledTimes(3);
  });

  it('should call onNodeFocus when a node is focused', () => {
    const handleNodeFocus = jest.fn();
    const treeNodes = buildTreeNodes();

    render(
      <Tree
        onNodeFocus={handleNodeFocus}
      >
        {ensureArray(treeNodes).map(node => (
          <TreeItemRender
            key={node.id}
            node={node}
          />
        ))}
      </Tree>
    );

    act(() => {
      screen.getByRole('tree').focus();
    });

    expect(handleNodeFocus).toHaveBeenCalledTimes(1);
    expect(handleNodeFocus).toHaveBeenCalledWith('1');
  });

  it('should call onNodeSelect when a node is selected', async () => {
    const user = userEvent.setup();
    const handleNodeSelect = jest.fn();
    const treeNodes = buildTreeNodes();

    render(
      <Tree
        isSelectable
        onNodeSelect={handleNodeSelect}
      >
        {ensureArray(treeNodes).map(node => (
          <TreeItemRender
            key={node.id}
            node={node}
          />
        ))}
      </Tree>
    );

    act(() => {
      screen.getByRole('tree').focus();
    });

    // nodeId="1"
    expect(screen.getByTestId('node-1')).not.toHaveAttribute('aria-selected');
    await user.click(screen.getByTestId('node-1-content'));
    expect(screen.getByTestId('node-1')).toHaveAttribute('aria-selected', 'true');

    // nodeId="7"
    expect(screen.getByTestId('node-7')).not.toHaveAttribute('aria-selected');
    await user.click(screen.getByTestId('node-7-content'));
    expect(screen.getByTestId('node-7')).toHaveAttribute('aria-selected', 'true');

    expect(handleNodeSelect).toHaveBeenCalledTimes(2);
    expect(handleNodeSelect).toHaveBeenNthCalledWith(1, ['1']);
    expect(handleNodeSelect).toHaveBeenNthCalledWith(2, ['7']);
  });

  it('should call onNodeToggle when a node is toggled', async () => {
    const user = userEvent.setup();
    const handleNodeToggle = jest.fn();
    const treeNodes = buildTreeNodes();

    render(
      <Tree
        onNodeToggle={handleNodeToggle}
      >
        {ensureArray(treeNodes).map(node => (
          <TreeItemRender
            key={node.id}
            node={node}
          />
        ))}
      </Tree>
    );

    act(() => {
      screen.getByRole('tree').focus();
    });

    // nodeId="1"
    expect(screen.getByTestId('node-1')).not.toHaveAttribute('aria-expanded');
    await user.click(screen.getByTestId('node-1-toggle'));
    expect(screen.getByTestId('node-1')).toHaveAttribute('aria-expanded', 'true');

    // nodeId="2"
    expect(screen.getByTestId('node-2')).not.toHaveAttribute('aria-expanded');
    await user.click(screen.getByTestId('node-2-toggle'));
    expect(screen.getByTestId('node-2')).toHaveAttribute('aria-expanded', 'true');

    expect(handleNodeToggle).toHaveBeenCalledTimes(2);
    expect(handleNodeToggle).toHaveBeenNthCalledWith(1, ['1']);
    expect(handleNodeToggle).toHaveBeenNthCalledWith(2, ['1', '2']);
  });

  it('should focus the first node if none of the nodes are selected before the tree receives focus', async () => {
    const treeNodes = buildTreeNodes();

    render(
      <Tree
        isSelectable
      >
        {ensureArray(treeNodes).map(node => (
          <TreeItemRender
            key={node.id}
            node={node}
          />
        ))}
      </Tree>
    );

    act(() => {
      screen.getByRole('tree').focus();
    });

    await waitFor(() => expect(screen.getByTestId('node-1-content')).toHaveFocus());
  });

  it('should focus the first selected node if some nodes are selected before the tree receives focus', async () => {
    const treeNodes = buildTreeNodes();
    const selectedNodeIds = ['7', '9'];

    render(
      <Tree
        isSelectable
        selected={selectedNodeIds}
      >
        {ensureArray(treeNodes).map(node => (
          <TreeItemRender
            key={node.id}
            node={node}
          />
        ))}
      </Tree>
    );

    act(() => {
      screen.getByRole('tree').focus();
    });

    await waitFor(() => expect(screen.getByTestId('node-7-content')).toHaveFocus());
  });

  it('should expand or collapse node when `Enter` key is pressed', async () => {
    const user = userEvent.setup();
    const handleKeyDown = jest.fn();
    const treeNodes = buildTreeNodes();

    render(
      <Tree
        onKeyDown={handleKeyDown}
      >
        {ensureArray(treeNodes).map(node => (
          <TreeItemRender
            key={node.id}
            node={node}
          />
        ))}
      </Tree>
    );

    act(() => {
      screen.getByRole('tree').focus();
    });

    expect(screen.getByTestId('node-1')).not.toHaveAttribute('aria-expanded');

    await user.keyboard('[Enter]');
    expect(screen.getByTestId('node-1')).toHaveAttribute('aria-expanded', 'true');

    await user.keyboard('[Enter]');
    expect(screen.getByTestId('node-1')).not.toHaveAttribute('aria-expanded');
  });

  it('should select or unselect node when `Space` key is pressed', async () => {
    const user = userEvent.setup();
    const handleKeyDown = jest.fn();
    const treeNodes = buildTreeNodes();

    render(
      <Tree
        isSelectable
        isUnselectable
        onKeyDown={handleKeyDown}
      >
        {ensureArray(treeNodes).map(node => (
          <TreeItemRender
            key={node.id}
            node={node}
          />
        ))}
      </Tree>
    );

    act(() => {
      screen.getByRole('tree').focus();
    });

    expect(screen.getByTestId('node-1')).not.toHaveAttribute('aria-selected');

    await user.keyboard('[Space]');
    expect(screen.getByTestId('node-1')).toHaveAttribute('aria-selected', 'true');

    await user.keyboard('[Space]');
    expect(screen.getByTestId('node-1')).not.toHaveAttribute('aria-selected');
  });

  it('should navigate the tree with specific keys', async () => {
    const user = userEvent.setup();
    const handleKeyDown = jest.fn();
    const treeNodes = buildTreeNodes();

    render(
      <Tree
        isSelectable
        isUnselectable
        multiSelect
        onKeyDown={handleKeyDown}
      >
        {ensureArray(treeNodes).map(node => (
          <TreeItemRender
            key={node.id}
            node={node}
          />
        ))}
      </Tree>
    );

    act(() => {
      screen.getByRole('tree').focus();
    });

    await user.keyboard('[Space]'); // Select node-1
    await user.keyboard('[ArrowRight]'); // Expand node-1
    await user.keyboard('[ArrowDown]'); // Focus node-2
    await user.keyboard('[Space]'); // Select node-2
    await user.keyboard('[ArrowRight]'); // Expand node-2
    await user.keyboard('[ArrowDown]'); // Focus node-3
    await user.keyboard('[Space]'); // Select node-3
    await user.keyboard('[ArrowRight]'); // Expand node-3
    await user.keyboard('[ArrowDown]'); // Focus node-4
    await user.keyboard('[Space]'); // Select node-4
    await user.keyboard('[ArrowUp]'); // Focus node-3
    await user.keyboard('[ArrowLeft]'); // Collapse node-3

    expect(screen.getByTestId('node-1')).toHaveAttribute('aria-selected', 'true');
    expect(screen.getByTestId('node-2')).toHaveAttribute('aria-selected', 'true');
    expect(screen.getByTestId('node-3')).toHaveAttribute('aria-selected', 'true');
    expect(screen.getByTestId('node-4')).toHaveAttribute('aria-selected', 'true');

    expect(screen.getByTestId('node-1')).toHaveAttribute('aria-expanded', 'true');
    expect(screen.getByTestId('node-2')).toHaveAttribute('aria-expanded', 'true');
    expect(screen.getByTestId('node-3')).not.toHaveAttribute('aria-expanded');

    await user.keyboard('[Home]'); // Focus node-1-content
    await waitFor(() => expect(screen.getByTestId('node-1-content')).toHaveFocus());

    await user.keyboard('[End]'); // Focus node-9-content
    await waitFor(() => expect(screen.getByTestId('node-9-content')).toHaveFocus());

    expect(handleKeyDown).toHaveBeenCalledTimes(14);
  });

  it('should select all nodes when `Ctrl+a` is pressed', async () => {
    const user = userEvent.setup();
    const handleNodeSelect = jest.fn();
    const treeNodes = buildTreeNodes();
    const treeMap = buildTreeMap(treeNodes);
    const expandableNodeIds = findExpandableNodeIds(treeNodes);
    const allNodeIds = Array.from(treeMap.keys());

    render(
      <Tree
        expanded={expandableNodeIds}
        isSelectable
        isUnselectable
        multiSelect
        onNodeSelect={handleNodeSelect}
      >
        {ensureArray(treeNodes).map(node => (
          <TreeItemRender
            key={node.id}
            node={node}
          />
        ))}
      </Tree>
    );

    act(() => {
      screen.getByRole('tree').focus();
    });

    await user.keyboard('{Control>}a{/Control}'); // translates to: Ctrl(down), a, Ctrl(up)
    expect(handleNodeSelect).toHaveBeenCalledTimes(1);
    expect(handleNodeSelect).toHaveBeenLastCalledWith(allNodeIds);
  });

  it('should allow selecting a range of nodes by holding down the `Ctrl` or `Shift` key while clicking with the mouse', async () => {
    const user = userEvent.setup();
    const handleNodeSelect = jest.fn();
    const treeNodes = buildTreeNodes();
    const expandableNodeIds = findExpandableNodeIds(treeNodes);

    render(
      <Tree
        expanded={expandableNodeIds}
        isSelectable
        isUnselectable
        multiSelect
        onNodeSelect={handleNodeSelect}
      >
        {ensureArray(treeNodes).map(node => (
          <TreeItemRender
            key={node.id}
            node={node}
          />
        ))}
      </Tree>
    );

    act(() => {
      screen.getByRole('tree').focus();
    });

    await user.keyboard('{Control>}'); // Press Control key without releasing it
    await user.click(screen.getByTestId('node-1-content'));
    await user.click(screen.getByTestId('node-2-content'));
    await user.click(screen.getByTestId('node-4-content'));
    await user.keyboard('{/Control}'); // Release Control key
    expect(handleNodeSelect).toHaveBeenLastCalledWith(['1', '2', '4']);

    await user.keyboard('{Shift>}'); // Press Shift key without releasing it
    await user.click(screen.getByTestId('node-8-content'));
    await user.keyboard('{/Shift}'); // Release Shift key
    expect(handleNodeSelect).toHaveBeenLastCalledWith(['1', '2', '4', '5', '6', '7', '8']);

    await user.keyboard('{Control>}'); // Press Control key without releasing it
    await user.click(screen.getByTestId('node-6-content'));
    await user.click(screen.getByTestId('node-7-content'));
    await user.keyboard('{/Control}'); // Release Control key
    expect(handleNodeSelect).toHaveBeenLastCalledWith(['1', '2', '4', '5', '8']);

    expect(handleNodeSelect).toHaveBeenCalledTimes(6);
  });

  it('should perform range selection from the last selected node to the first node using the `Ctrl+Shift+Home` shortcut', async () => {
    const user = userEvent.setup();
    const handleNodeSelect = jest.fn();
    const treeNodes = buildTreeNodes();
    const expandableNodeIds = findExpandableNodeIds(treeNodes);

    render(
      <Tree
        expanded={expandableNodeIds}
        isSelectable
        isUnselectable
        multiSelect
        onNodeSelect={handleNodeSelect}
      >
        {ensureArray(treeNodes).map(node => (
          <TreeItemRender
            key={node.id}
            node={node}
          />
        ))}
      </Tree>
    );

    act(() => {
      screen.getByRole('tree').focus();
    });

    await user.click(screen.getByTestId('node-9-content')); // the second to last node
    expect(handleNodeSelect).toHaveBeenLastCalledWith(['9']);

    await user.keyboard('{Control>}{Shift>}{Home}{/Shift}{/Control}');
    expect(handleNodeSelect).toHaveBeenLastCalledWith(['9', '1', '2', '3', '4', '5', '6', '7', '8']);
  });

  it('should perform range selection from the last selected node to the last node using the `Ctrl+Shift+End` shortcut', async () => {
    const user = userEvent.setup();
    const handleNodeSelect = jest.fn();
    const treeNodes = buildTreeNodes();
    const expandableNodeIds = findExpandableNodeIds(treeNodes);

    render(
      <Tree
        expanded={expandableNodeIds}
        isSelectable
        isUnselectable
        multiSelect
        onNodeSelect={handleNodeSelect}
      >
        {ensureArray(treeNodes).map(node => (
          <TreeItemRender
            key={node.id}
            node={node}
          />
        ))}
      </Tree>
    );

    act(() => {
      screen.getByRole('tree').focus();
    });

    await user.click(screen.getByTestId('node-2-content')); // the second node
    expect(handleNodeSelect).toHaveBeenLastCalledWith(['2']);

    await user.keyboard('{Control>}{Shift>}{End}{/Shift}{/Control}');
    expect(handleNodeSelect).toHaveBeenLastCalledWith(['2', '3', '4', '5', '6', '7', '8', '9', '10']);
  });
});
