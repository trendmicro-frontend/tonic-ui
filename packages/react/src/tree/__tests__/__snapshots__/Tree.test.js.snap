// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree should render correctly 1`] = `
.emotion-0 {
  outline: 0;
}

.emotion-4 {
  color: var(--tonic-colors-white-primary);
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  outline: 0;
  padding-left: var(--tonic-space-3x);
  padding-right: var(--tonic-space-3x);
  padding-top: var(--tonic-space-2x);
  padding-bottom: var(--tonic-space-2x);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 100%;
}

.emotion-4:focus-visible {
  outline-color: var(--tonic-colors-blue-60);
  outline-offset: calc(var(--tonic-space-1h) * -1);
  outline-style: solid;
  outline-width: var(--tonic-sizes-1h);
}

.emotion-4:hover,
.emotion-4[data-hover] {
  background-color: rgba(255, 255, 255, 0.12);
}

.emotion-4:hover+[role="group"] {
  position: relative;
}

.emotion-4:hover+[role="group"]::before {
  background-color: rgba(255, 255, 255, 0.12);
  content: "";
  position: absolute;
  top: 0;
  bottom: 0;
  left: 19.5px;
  width: 1px;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  width: var(--tonic-sizes-6x);
}

.emotion-8 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: var(--tonic-colors-transparent);
  border: none;
  color: inherit;
  cursor: pointer;
  outline: 0;
  padding: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-10 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  color: var(--tonic-colors-white-secondary);
  -webkit-transform: rotate(0deg);
  -moz-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
  -webkit-transition: -webkit-transform 93ms cubic-bezier(0.0, 0, 0.2, 1);
  transition: transform 93ms cubic-bezier(0.0, 0, 0.2, 1);
}

.emotion-10:hover,
.emotion-10[data-hover] {
  color: var(--tonic-colors-white-primary);
}

.emotion-12 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: var(--tonic-sizes-4x);
  height: var(--tonic-sizes-4x);
}

.emotion-14 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
  font-weight: var(--tonic-fontWeights-normal);
}

<div>
  <div
    aria-label="tree"
    aria-multiselectable="true"
    class="emotion-0 emotion-1"
    id="4"
    role="tree"
    tabindex="0"
  >
    <div
      class="emotion-2 emotion-1"
      data-testid="node-1"
      id="4-1"
      role="treeitem"
    >
      <div
        class="emotion-4 emotion-1"
        data-testid="node-1-content"
        tabindex="-1"
      >
        <div
          class="emotion-6 emotion-1"
        >
          <button
            aria-label="toggle"
            class="emotion-8 emotion-1"
            data-testid="node-1-toggle"
            role="button"
            tabindex="0"
            type="button"
          >
            <div
              class="emotion-10 emotion-1"
            >
              <svg
                aria-hidden="true"
                class="emotion-12 emotion-1"
                fill="currentColor"
                focusable="false"
                viewBox="0 0 16 16"
              >
                <g>
                  <path
                    d="M6.501 12.5l-1-1 3.501-3.506-3.49-3.494 1-1 4.488 4.494z"
                  />
                </g>
              </svg>
            </div>
          </button>
        </div>
        <div
          class="emotion-14 emotion-1"
          id="@tonic-ui/react:TooltipTrigger-1"
        >
          Node 1
        </div>
      </div>
    </div>
    <div
      class="emotion-2 emotion-1"
      data-testid="node-7"
      id="4-7"
      role="treeitem"
    >
      <div
        class="emotion-4 emotion-1"
        data-testid="node-7-content"
        tabindex="-1"
      >
        <div
          class="emotion-6 emotion-1"
        >
          <button
            aria-label="toggle"
            class="emotion-8 emotion-1"
            data-testid="node-7-toggle"
            role="button"
            tabindex="0"
            type="button"
          >
            <div
              class="emotion-10 emotion-1"
            >
              <svg
                aria-hidden="true"
                class="emotion-12 emotion-1"
                fill="currentColor"
                focusable="false"
                viewBox="0 0 16 16"
              >
                <g>
                  <path
                    d="M6.501 12.5l-1-1 3.501-3.506-3.49-3.494 1-1 4.488 4.494z"
                  />
                </g>
              </svg>
            </div>
          </button>
        </div>
        <div
          class="emotion-14 emotion-1"
          id="@tonic-ui/react:TooltipTrigger-2"
        >
          Node 7
        </div>
      </div>
    </div>
    <div
      class="emotion-2 emotion-1"
      data-testid="node-9"
      id="4-9"
      role="treeitem"
    >
      <div
        class="emotion-4 emotion-1"
        data-testid="node-9-content"
        tabindex="-1"
      >
        <div
          class="emotion-6 emotion-1"
        >
          <button
            aria-label="toggle"
            class="emotion-8 emotion-1"
            data-testid="node-9-toggle"
            role="button"
            tabindex="0"
            type="button"
          >
            <div
              class="emotion-10 emotion-1"
            >
              <svg
                aria-hidden="true"
                class="emotion-12 emotion-1"
                fill="currentColor"
                focusable="false"
                viewBox="0 0 16 16"
              >
                <g>
                  <path
                    d="M6.501 12.5l-1-1 3.501-3.506-3.49-3.494 1-1 4.488 4.494z"
                  />
                </g>
              </svg>
            </div>
          </button>
        </div>
        <div
          class="emotion-14 emotion-1"
          id="@tonic-ui/react:TooltipTrigger-3"
        >
          Node 9
        </div>
      </div>
    </div>
  </div>
</div>
`;
